File: ios/Packages/SimplicityApi/Sources/SimplicityApi/ApiAdapter.swift
import Foundation

/// What this package needs but cannot know for itself: where the API is, and who is calling.
///
/// The app supplies it. Keeping it a protocol is what lets the package be tested without an app,
/// and what keeps Cognito out of it entirely — nothing here knows how a token is obtained, only
/// that asking for one may take a moment and may come back empty.
public protocol ApiAdapter: Sendable {
    var baseURL: URL { get }
    func accessToken() async -> String?

    /// A token obtained by forcing a refresh, rather than whatever is cached.
    ///
    /// Needed because the server can now void a token before it expires — being removed from an
    /// organisation does exactly that. The cached token would keep failing until it aged out, so a
    /// rejected request has to be able to insist on a new one.
    func refreshedAccessToken() async -> String?
}

File: android/api/src/main/kotlin/io/simplicity/training/api/ApiAdapter.kt
package io.simplicity.training.api

/**
 * What `:api` cannot know for itself. The app supplies it.
 *
 * Keeping it an interface is what lets this module be tested without an app, and what keeps
 * Cognito out of it entirely — nothing here knows how a token is obtained, only that asking for one
 * may take a moment and may come back empty.
 */
interface ApiAdapter {

    val baseUrl: String

    /** Whatever is cached. Null when signed out, which is not an error: some endpoints are public. */
    suspend fun accessToken(): String?

    /**
     * A token obtained by forcing a refresh, for a request the server has just rejected.
     *
     * Distinct from [accessToken], which returns what is held and only refreshes near expiry. The
     * server can void a token long before then — removal from an organisation does exactly that —
     * and only insisting on a new one recovers from it.
     */
    suspend fun refreshedAccessToken(): String?
}
